implementation-details.md

```diff
@@ -22,16 +22,26 @@ Each Solana node participates in gossip in one of two modes:
+Before the Gossip service is started, the node must discover the shred version and check own ports using the entrypoint's IP echo server listening on TCP endpoint on the gossip address.
+The node sends a IP echo server request with a list of UDP and TCP ports whose availability needs to be checked by the entrypoint node.
+The entrypoint node tries to establish a connection for TCP ports and send packets for UDP ports and then the entrypoint node replies back with the IP echo server response containing the public address of the node and a shred version.
+
+The node starts the Gossip service and then advertises itself to the cluster by sending messages to the peer identified by the entrypoint. The node sends:
+* two [push messages](#push-messages) containing its own [`Version`](gossip-protocol-spec.md#version) and [`NodeInstance`](gossip-protocol-spec.md#nodeinstance),
+* a [pull request](#pull-requests) that contains the node's own [`LegacyContactInfo`](gossip-protocol-spec.md#legacycontactinfo).
+
+The push messages are [propagated further](#propagating-push-messages-over-a-cluster) to other nodes in the cluster. Soon after that, the node should start receiving [pull responses](#pull-responses), push messages, and [pings](#ping-and-pong-messages) from other nodes.
 
 When the connection with the cluster is established, the node can fully participate in the gossip - it can update the cluster with its data (push messages) and receive data from its peers through received push messages and by sending pull requests and receiving the corresponding pull responses.
 
+However, for the node to switch from the default `spy` mode to the `gossip` mode, it awaits a `SnapshotHashes` message from the cluster.
+Once it receives the message, the node will try to download the snapshots from any known RPC servers in the cluster.
+After the snapshot download, the node runs in the `gossip` mode.
+
 ### Keeping the connection alive
-To keep the connection with other nodes, nodes must constantly exchange messages. Nodes in the cluster send around 30-40 pull requests per second and push messages every 7.5 seconds. If a node stops sending pull requests and push messages for more than 15 seconds it will be ignored by other nodes and will not receive any push messages or pull requests. Sending ping messages only is not enough, node needs to send push messages or pull requests at least every 15 seconds to keep the connection alive.
+To keep the connection with other nodes, nodes must constantly exchange messages. Nodes in the cluster send around 10-20 pull requests per second and push messages every 7.5 seconds. If a node stops sending pull requests and push messages for more than 15 seconds it will be ignored by other nodes and will not receive any push messages or pull requests. Sending ping messages only is not enough, node needs to send push messages or pull requests at least every 15 seconds to keep the connection alive.
```

Comment:
I wouldn't state the actual number of pull requests sent per second. because it depends on the cluster and the difference between node's crds tables i believe. 

I would also say that within the PushMessages, you need to be sending your contact info and node instance data. I'm pretty sure you won't stay connected to the cluster if you just send `Votes` for example. since your contact info will eventually get purged from other node's crds tables

Reply:
> I wouldn't state the actual number of pull requests sent per second. because it depends on the cluster and the difference between node's crds tables i believe.

Would `Nodes in the cluster send around lots of pull requests per second and push messages every 7.5 seconds.`
Will this be better maybe?

Reply:
> I would also say that within the PushMessages, you need to be sending your contact info and node instance data. I'm pretty sure you won't stay connected to the cluster if you just send Votes for example. since your contact info will eventually get purged from other node's crds tables

I added extra context for this